formats/protobuf/commonTest/src/kotlinx/serialization/protobuf/ProtobufOneOfTest.kt

```diff
@@ -0,0 +1,464 @@
+        }
+    }
+
+    @Test
+    fun testOneOfDecodeNullable() {
+        ProtoBuf.decodeFromHexString<OneOfDataNullable>("12036261721a03666f6f").let {
+            assertEquals(OneOfDataNullable(StringType("bar"), "foo"), it)
+        }
+        ProtoBuf.decodeFromHexString<OneOfDataNullable>("082a1a03666f6f").let {
+            assertEquals(OneOfDataNullable(IntType(42), "foo"), it)
+        }
+        ProtoBuf.decodeFromHexString<OneOfDataNullable>("1a03666f6f").let {
+            assertEquals(OneOfDataNullable(null, "foo"), it)
+        }
+    }
+
+    @Test
+    fun testOneOfToFlatInt() {
+        val dataInt = OneOfData(IntType(42), "foo")
+        ProtoBuf.encodeToByteArray(dataInt).let {
+            ProtoBuf.decodeFromByteArray<DataNoOneOf>(it).let { data ->
+                assertEquals(
+                    DataNoOneOf(42, "", "foo"), data
+                )
+            }
+        }
+    }
+
+    @Test
+    fun testOneOfToFlatString() {
+        val dataString = OneOfData(StringType("bar"), "foo")
+        ProtoBuf.encodeToByteArray(dataString).let {
+            ProtoBuf.decodeFromByteArray<DataNoOneOf>(it).let { data ->
+                assertEquals(
+                    DataNoOneOf(0, "bar", "foo"), data
+                )
+            }
+        }
+    }
+
+    @Test
+    fun testFlatIntToOneOf() {
+        val dataInt = DataNoOneOf(42, "", "foo")
+        ProtoBuf.encodeToByteArray(dataInt).let {
+            ProtoBuf.decodeFromByteArray<OneOfData>(it).let { data ->
+                assertEquals(
+                    OneOfData(IntType(42), "foo"), data
+                )
+            }
+        }
+    }
+
+    @Test
+    fun testFlatStringToOneOf() {
+        val dataString = DataNoOneOf(0, "bar", "foo")
+        ProtoBuf.encodeToByteArray(dataString).let {
+            ProtoBuf.decodeFromByteArray<OneOfData>(it).let { data ->
+                assertEquals(
+                    OneOfData(StringType("bar"), "foo"), data
+                )
+            }
+        }
+    }
+
+    @Test
+    fun testEncodeNestedStruct() {
+        val data = OneOfData(NestedIntType(IntType(32)), "foo")
+        ProtoBuf.encodeToHexString(OneOfData.serializer(), data).also {
+            println(it)
+            /**
+             * 7: {5: 32}
+             * 3: {"foo"}
+             */
+            assertEquals("3a0228201a03666f6f", it)
+        }
+    }
+
+    @Test
+    fun testDecodeNestedStruct() {
+        val data = OneOfData(NestedIntType(IntType(32)), "foo")
+        ProtoBuf.decodeFromHexString<OneOfData>("3a0228201a03666f6f").also {
+            println(it)
+            assertEquals(data, it)
+        }
+    }
+
+    @Serializable
+    @ProtoNumber(8)
+    data class FailType(val i: Int, val j: Int) : IType
+
+    @Test
+    fun testOneOfElementCheck() {
+        val data = OneOfData(FailType(1, 2), "foo")
+        assertFailsWith<IllegalArgumentException> { ProtoBuf.encodeToHexString(OneOfData.serializer(), data) }
```

Comment:
Here and in other tests: It is better to also check that exception message contains some meaningful text which will help the user to debug a problem. (e.g. in this case, the message should contain 'Implementation of oneOf type ... should contain only 1 element'). You can copy `assertFailsWithMessage` from json module (https://github.com/Kotlin/kotlinx.serialization/blob/cf57414d83314bf05cb28ae8a600163b96855ba0/formats/json-tests/commonTest/src/kotlinx/serialization/test/TestingFramework.kt#L84) or write a different utility